### ISSUE TITLE
[Feature Request] How to recompute the advantage of a  minibatch in ppo

### ISSUE THREAD

**Issue by yangysc** [DLR-RM/stable-baselines3#445]
### 🚀 Feature

According to this [paper](https://arxiv.org/pdf/2006.05990), recomputing the advantage can be helpful for the PPO performance.
The function is provided by `tianshou` library.

https://github.com/thu-ml/tianshou/blob/655d5fb14fe85ea9da86b441456286fa1f078384/tianshou/policy/modelfree/ppo.py#L107

But I don't know how to add this in sb3. Some hints about how to do that would be very helpful.

Thanks!

### Motivation

I am comparing stable-baselines3, tianshou and rllib for the best performance of PPO.

### Pitch

Recompute the advantage in learning ppo.

### Checklist

- [X] I have checked that there is no similar [issue](https://github.com/DLR-RM/stable-baselines3/issues) in the repo (**required**)

**Comment by Miffyli** [comment]
Hmm indeed it seems to be a beneficial and simple trick to add, but is not part of the original PPO (not a "baseline"). Granted, we have some other small modifications in the PPO compared to the original OAI baselines. @araffin what would be your take? I am up for such a feature (not too intrusive, seems beneficial, has proper experiments behind it).

**Comment by yangysc** [comment]
> Hmm indeed it seems to be a beneficial and simple trick to add, but is not part of the original PPO (not a "baseline"). Granted, we have some other small modifications in the PPO compared to the original OAI baselines. @araffin what would be your take? I am up for such a feature (not too intrusive, seems beneficial, has proper experiments behind it).

Thanks for your support. Yes, it does not belong to the original PPO, but maybe we can leave it as a option and let more users give feedbacks.  I would like to benchmark the difference on my personal task if it comes out.

**Comment by araffin** [comment]
Hello,
if you want to try it out, you mostly need to avoid overwriting the variables here: https://github.com/DLR-RM/stable-baselines3/blob/master/stable_baselines3/common/buffers.py#L441
and recompute values (using current vf function) + advantages (using `compute_returns_and_advantage()`) at every epoch.


On including that feature into SB3, I'm afraid that would require to hack the code in several places...

**Comment by yangysc** [comment]
> compute_returns_and_advantage

Hello, many thanks for your help. 
The second parameter for `compute_returns_and_advantage()` is the variable `done`. How can I get this from the buffer?   
If it is not very trouble for changing it only for `ppo` for a little try and test, it would be very kind of you for showing a small code. I just worry for making any mistake, so the agent would learn for no reason. Sorry for bothering you.
Thanks again.

**Comment by araffin** [comment]
the last episode starts are saved: https://github.com/DLR-RM/stable-baselines3/blob/df6f9de8f46509dad47e6d2e5620aa993b0fc883/stable_baselines3/common/on_policy_algorithm.py#L200
and the last values can be recomputed using last observation.

Please give it a try and then you can put a link to your code if we want to checkout ;)

**Comment by yangysc** [comment]
> the last episode starts are saved:
> 
> https://github.com/DLR-RM/stable-baselines3/blob/df6f9de8f46509dad47e6d2e5620aa993b0fc883/stable_baselines3/common/on_policy_algorithm.py#L200
> 
> 
> and the last values can be recomputed using last observation.
> Please give it a try and then you can put a link to your code if we want to checkout ;)

Thanks for your prompt.  I have uploaded the modified files [here](https://github.com/yangysc/recompute_adv/commit/016249799added4c16f606708495129da42b1f44). However, it wouldn't work even on CartPole....
Would you mind having a  look?

Specifically, the modified files are
1. type_aliases.py


https://github.com/yangysc/stable_baselines_recompute_advantage/blob/712815de60924cf6f47150074e24d8251ff56c38/stable-baselines3/stable_baselines3/common/type_aliases.py#L31-L35
 
2. buffers.py
https://github.com/yangysc/stable_baselines_recompute_advantage/blob/712815de60924cf6f47150074e24d8251ff56c38/stable-baselines3/stable_baselines3/common/buffers.py#L461-L467


3. ppo.py
https://github.com/yangysc/recompute_adv/blob/016249799added4c16f606708495129da42b1f44/stable-baselines3/stable_baselines3/ppo/ppo.py#L196-216

https://github.com/yangysc/recompute_adv/blob/016249799added4c16f606708495129da42b1f44/stable-baselines3/stable_baselines3/ppo/ppo.py#L327-368

If we do not recompute the advantage, the Cartpole example result is 
```
-----------------------------------------
| rollout/                |             |
|    ep_len_mean          | 58.1        |
|    ep_rew_mean          | 58.1        |
| time/                   |             |
|    fps                  | 1309        |
|    iterations           | 5           |
|    time_elapsed         | 7           |
|    total_timesteps      | 10240       |
| train/                  |             |
|    approx_kl            | 0.008776829 |
|    clip_fraction        | 0.0782      |
|    clip_range           | 0.2         |
|    entropy_loss         | -0.608      |
|    explained_variance   | 0.326       |
|    learning_rate        | 0.0003      |
|    loss                 | 25.1        |
|    n_updates            | 40          |
|    policy_gradient_loss | -0.0177     |
|    value_loss           | 54.5        |
```
If we recompute like I did,  the result is much worse,
```
-------------------------------------------
| rollout/                |               |
|    ep_len_mean          | 10.3          |
|    ep_rew_mean          | 10.3          |
| time/                   |               |
|    fps                  | 865           |
|    iterations           | 49            |
|    time_elapsed         | 116           |
|    total_timesteps      | 100352        |
| train/                  |               |
|    approx_kl            | 0.00023333642 |
|    clip_fraction        | 0.00283       |
|    clip_range           | 0.2           |
|    entropy_loss         | -0.216        |
|    explained_variance   | 0.943         |
|    learning_rate        | 0.0003        |
|    loss                 | 0.254         |
|    n_updates            | 480           |
|    policy_gradient_loss | 0.000201      |
|    value_loss           | 0.494         |
-------------------------------------------

```
Thank you !

**Comment by yangysc** [comment]
> 
> 
> Hello,
> if you want to try it out, you mostly need to avoid overwriting the variables here: https://github.com/DLR-RM/stable-baselines3/blob/master/stable_baselines3/common/buffers.py#L441
> and recompute values (using current vf function) + advantages (using `compute_returns_and_advantage()`) at every epoch.
> 
> On including that feature into SB3, I'm afraid that would require to hack the code in several places...

By the way, I did not avoid overwriting the variables here: https://github.com/DLR-RM/stable-baselines3/blob/master/stable_baselines3/common/buffers.py#L441.  Why we need to do this? Could this be an issue?

**Comment by yangysc** [comment]
solved.......:)  Thanks

**Comment by davidsblom** [comment]
@yangysc does it help to recompute the advantage?

**Comment by yangysc** [comment]
> @yangysc does it help to recompute the advantage?
I did as `araffin` kindly suggested. Here is the result for Cartpole. It seems recomputing would be slightly better for this case, despite the `fps` would drop.
<img width="1623" alt="recompute_adv_cartpole" src="https://user-images.githubusercontent.com/8980981/119311586-d722b800-bca3-11eb-8316-1f48c0f052c2.png">

**Comment by araffin** [comment]
thanks for sharing, btw, are you recomputing it after each epoch or each gradient step?
and is there a public link (if someone else is interested in the future) ?

**Comment by yangysc** [comment]
Hello, I recompute it after each epoch, being consistent with the library `tianshou`
https://github.com/thu-ml/tianshou/blob/655d5fb14fe85ea9da86b441456286fa1f078384/tianshou/policy/modelfree/ppo.py#L107

I would like to upload the modified ppo later, and hopefully you can help check if there is any potential problems.

To support multi-envs, I did what you suggested before, avoid overwriting the variables https://github.com/DLR-RM/stable-baselines3/blob/master/stable_baselines3/common/buffers.py#L441, and reshape them whenever sampling.  We  don't have to do this if we only use one env. But it heavilly slows low the learning process...
![Uploading image.png…]()

```
    def _get_samples(self, batch_inds: np.ndarray, env: Optional[VecNormalize] = None) -> RolloutBufferSamples:
        data = (
            self.swap_and_flatten(self.observations)[batch_inds],
            self.swap_and_flatten(self.actions)[batch_inds],
            self.swap_and_flatten(self.values[batch_inds]).flatten(),
            self.swap_and_flatten(self.log_probs[batch_inds]).flatten(),
            self.swap_and_flatten(self.advantages[batch_inds]).flatten(),
            self.swap_and_flatten(self.returns[batch_inds]).flatten(),
            # self.swap_and_flatten(self.to_torch(self.rewards[batch_inds]).flatten()),
            # self.swap_and_flatten(self.to_torch(self.episode_starts[batch_inds]).flatten()),
            # batch_inds

        )
        return RolloutBufferSamples(*tuple(map(self.to_torch, data)))
```
I tried recoputing it after sampling `bs` observations so this is each gradient step yesterday. But unluckily, I didn't sucessfully make it :<

**Comment by araffin** [comment]
>I pasted the main modification below. Hopefully you can help check if there is any potential problems.

I see that you are re-computing the last values but not updating the values in the rollout buffer, is that intended?

**Comment by yangysc** [comment]
> I see that you are re-computing the last values but not updating the values in the rollout buffer, is that intended?
Fixed this.
```python
   def train(self) -> None:
        """
        Update policy using the currently gathered rollout buffer.
        """
        # Update optimizer learning rate
        self._update_learning_rate(self.policy.optimizer)
        # Compute current clip range
        clip_range = self.clip_range(self._current_progress_remaining)
        # Optional: clip range for the value function
        if self.clip_range_vf is not None:
            clip_range_vf = self.clip_range_vf(self._current_progress_remaining)

        entropy_losses = []
        pg_losses, value_losses = [], []
        clip_fractions = []

        continue_training = True

        # train for n_epochs epochs
        for epoch in range(self.n_epochs):
            approx_kl_divs = []
            # Do a complete pass on the rollout buffer
            
            # recomput adv
            with th.no_grad():
                _, last_values, _ = self.policy(obs_as_tensor(self._last_obs, self.device))
                _, new_values, _ = self.policy(obs_as_tensor(self.rollout_buffer.observations , self.device))
                self.rollout_buffer.values = new_values.detach().numpy()
                self.rollout_buffer.compute_returns_and_advantage(last_values, dones=self._last_episode_starts)
                
            for rollout_data in self.rollout_buffer.get(self.batch_size):
                actions = rollout_data.actions
                if isinstance(self.action_space, spaces.Discrete):
                    # Convert discrete action from float to long
                    actions = rollout_data.actions.long().flatten()

                # Re-sample the noise matrix because the log_std has changed
                # TODO: investigate why there is no issue with the gradient
                # if that line is commented (as in SAC)
                if self.use_sde:
                    self.policy.reset_noise(self.batch_size)

                values, log_prob, entropy = self.policy.evaluate_actions(rollout_data.observations, actions)
                values = values.flatten()

                # indices = rollout_data.indices
                # reward = self.rollout_buffer.rewards[indices].flatten()
                #
                # episode_starts = rollout_data.episode_starts
                #
                # # indices = rollout_data[6]  # this is the pos we append indice to DictRolloutBufferSamples, see L735 of buffers.py
                # # values, log_prob, entropy = self.policy.evaluate_actions(rollout_data.observations, actions)
                # with th.no_grad():
                #     # Recompute value for the last timestep
                #     _, last_values, _ = self.policy(obs_as_tensor(self._last_obs, self.device))
                # returns, advantages = compute_returns_and_advantage(last_values=last_values,
                #                                            rewards=reward,
                #                                            values=values.detach(),
                #                                            episode_starts=episode_starts,
                #                                            gamma=self.gamma,
                #                                            gae_lambda=self.gae_lambda,
                #                                            buffer_size=self.batch_size,
                #                                            dones=self._last_episode_starts)
                # Normalize advantage
                advantages = rollout_data.advantages
```
![image](https://user-images.githubusercontent.com/8980981/119327147-3dfc9d00-bcb5-11eb-85b9-b0e25b1ffe8f.png)



It seems recomputing is slightly better, with the cost of lower fps.

**Comment by araffin** [comment]
>It seems recomputing is slightly better, with the cost of lower fps. The entropy loss seems to be much lower.

Usually CartPole is nice to debug and check you did not break anything but overall I would go for a more challenging environment like `HalfCheetahBulletEnv-v0` (or `AntBullet`, `Hopper`  and the harder `Walker2DBulletEnv-v0`) to see if any improvement can be achieved. I would recommend to use the [rl zoo](https://github.com/DLR-RM/rl-baselines3-zoo) for such comparison (the learning curve for normal PPO is already there).

**Comment by yangysc** [comment]
I agree. If the code above is fine, I would like to compare them later

**Comment by araffin** [comment]
>I agree. If the code above is fine, I would like to compare them later

Looks ok but I would double check the shapes

**Comment by yangysc** [comment]
> > I agree. If the code above is fine, I would like to compare them later
> 
> Looks ok but I would double check the shapes

Cool, thanks again :)

**Comment by Karlheinzniebuhr** [comment]
> I agree. If the code above is fine, I would like to compare them later

Hi! Where you able to test your implementation on more environments? Would love to know how it performed
